### DISCUSSION TITLE
software reset

### DISCUSSION THREAD

**Discussion by mauroForlimpopoli** [Moddable-OpenSource/moddable#579] [Q&A] [answered]
Is there a "software restar command" for the Moddable Two? (like reset button)

**Comment by phoddie** [comment, marked as answer]
There is a restart function available as part of some examples. There is not, however, a standalone restart function. We should do that, and the [Host Provider Instance](https://ecmatc53.github.io/spec/web/spec.html#-16-host-provider-instance) in our work with Ecma TC53 is probably a good place for it.

Meanwhile, it is easy to add yourself. The details are explained step-by-step in our book, [IoT Development for ESP32 and ESP8266 with JavaScript](https://www.moddable.com/book), in the "Adding Native Code" chapter. Restart is very much like the random integer example on pages 521-523..

I'll just give a quick summary here. You need to add one file to your project folder. Name it `restart.c`. It just contains a native function to  invoke the restart function in the ESP-IDF:

```c
#include "xs.h"

void xs_restart(xsMachine *the)
{
	esp_restart();
}
```
Then in your project, declare a JavaScript function to connect to the native function (the `@` syntax defines the connection). From there, you call it like any other JavaScript function.

```js
function restart() @ "xs_restart"

restart();
```

To build  `restart.c`, you need to add the file to your manifest. Here's how that looks if you add it to [$MODDABLE/examples/helloworld/manifest.json](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/helloworld/manifest.json):

```json
{
	"include": "$(MODDABLE)/examples/manifest_base.json",
	"modules": {
		"*": [
			"./main",
			"./restart"
		]
	}
}
```

Good luck!

**Reply by mauroForlimpopoli** [reply]
Thank  you Peter, it works perfectly !
